CLW/CLWParallelPrimitives.cpp

```diff
@@ -921,65 +920,181 @@ void CLWParallelPrimitives::ReclaimTempBuffer(std::map<size_t, CLWBuffer<T> > co
     collection[buffer.GetElementCount()] = buffer;
 }
 
+CLWEvent CLWParallelPrimitives::Copy(unsigned int deviceIdx, CLWBuffer<cl_int> input, CLWBuffer<cl_int> output, int numElems)
+{
+    int ELEMS_PER_WI = 4;
+    int GROUP_BLOCK_SIZE = (WG_SIZE * ELEMS_PER_WI);
+    int NUM_BLOCKS = (numElems + GROUP_BLOCK_SIZE - 1) / GROUP_BLOCK_SIZE;
+
+    CLWKernel copyKernel = program_.GetKernel("copy");
+
+    copyKernel.SetArg(0, input);
+    copyKernel.SetArg(1, numElems);
+    copyKernel.SetArg(2, output);
+
+    return context_.Launch1D(0, NUM_BLOCKS * WG_SIZE, WG_SIZE, copyKernel);
+}
+
+const float epsilon = 0.001f;
+
 template <class T>
-CLWEvent CLWParallelPrimitives::Normalize(const char* kernelName, unsigned int deviceIdx, CLWBuffer<T> input, CLWBuffer<T> output, int numElems, int groupSize, CLWBuffer<T> cache)
+T CLWParallelPrimitives::GetMaxNum()
+{
+    return std::numeric_limits<T>::max();
+}
+
+
+template <>
```

Comment:
Why do we need this functions in CLWPrimitives?

Reply:
Just helpers, cause cl_float3